src/Fixer/Preload/PreloadExplicitClassSymbolsFixer.php

```diff
@@ -0,0 +1,282 @@
+        return $tokens->isTokenKindFound(T_CLASS) || $tokens->isTokenKindFound(T_TRAIT);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        $this->tokenAnalyzer = new TokensAnalyzer($tokens);
+        $candidates = $this->parse($tokens, '__construct');
+        if (empty($candidates)) {
+            return;
+        }
+
+        $candidates = array_unique($candidates);
+        $classesNotToLoad = $this->getPreloadedClasses($tokens);
+
+        $classesToLoad = array_diff($candidates, $classesNotToLoad);
+        $this->injectClasses($tokens, $classesToLoad);
+    }
+
+    /**
+     * Parse functions for used classes. Make a recursive call to other function.
+     * @param string $functionName
+     *
+     * @return string[] classes
+     */
+    private function parse(Tokens $tokens, $functionName)
+    {
+        if (null === $functionIndex = $this->findFunction($tokens, $functionName)) {
+            return [];
+        }
+
+        $classes = [];
+        $methodAttributes = $this->tokenAnalyzer->getMethodAttributes($functionIndex);
+
+        // If not public
+        if (T_PRIVATE === $methodAttributes['visibility'] || T_PROTECTED === $methodAttributes['visibility']) {
+            // Get argument types
+            $arguments = $this->functionsAnalyzer->getFunctionArguments($tokens, $functionIndex);
+            foreach ($arguments as $argument) {
+                if ($argument->hasTypeAnalysis() && !$argument->getTypeAnalysis()->isReservedType()) {
+                    $classes[] = $argument->getTypeAnalysis()->getName();
+                }
+            }
+
+            // Get return type
+            $returnType = $this->functionsAnalyzer->getFunctionReturnType($tokens, $functionIndex);
+            if (null !== $returnType && !$returnType->isReservedType()) {
+                $classes[] = $returnType->getName();
+            }
+        }
+
+        // Parse the body of the method
+        $blockStart = $tokens->getNextTokenOfKind($functionIndex, ['{']);
+        $blockEnd = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_CURLY_BRACE, $blockStart);
+
+        for ($i = $blockStart; $i < $blockEnd; ++$i) {
+            $token = $tokens[$i];
+            // Find Foo::class, new Foo() and function calls.
+            if ($token->isGivenKind(T_NEW)) {
+                // FIXME We need to support classes like \Biz\BazClass
+                $class = $tokens[$tokens->getNextMeaningfulToken($i)]->getContent();
+                $classes[] = $class;
+            } elseif ($token->isGivenKind(T_DOUBLE_COLON)) {
+                // FIXME We need to support classes like \Biz\BazClass
+                $class = $tokens[$tokens->getPrevMeaningfulToken($i)]->getContent();
+                $classes[] = $class;
+            } elseif ($token->isGivenKind(T_OBJECT_OPERATOR)) {
+                // FIXME Better check if function call to avoid false positive like "$this->bar = 2;"
```

Comment:
@Nyholm , are those 3 `FIXME` cases covered by tests already? if not, maybe we can add a test case for each, so we can easier understand what is example code snippet you struggle with and expected outcome ?

Reply:
They are. They are the failing tests. 
I'll work with them now to see if I can make sure they are green

Reply:
Tests are green. Locally at least. =)